File: apps/backends/rails/app/controllers/products_controller.rb
class ProductsController < ApplicationController
  def index
    cols = %i[id sku name price_cents stock]
    rows = Product.order(:name).pluck(*cols)
    products = rows.map { |r| cols.zip(r).to_h }
    render json: { products: products }
  end

  def show
    product = Product.find_by(id: params[:id])
    if product.nil?
      render json: { error: 'not found' }, status: :not_found
      return
    end
    render json: { id: product.id, sku: product.sku, name: product.name, price_cents: product.price_cents, stock: product.stock }
  end
end

File: apps/backends/php/app/Http/Controllers/ProductsController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\JsonResponse;
use Illuminate\Support\Facades\DB;

class ProductsController extends Controller
{
    public function __invoke(): JsonResponse
    {
        $rows = DB::select('SELECT id, sku, name, price_cents, stock FROM products ORDER BY name');

        $products = array_map(fn($r) => [
            'id' => $r->id,
            'sku' => $r->sku,
            'name' => $r->name,
            'price_cents' => $r->price_cents,
            'stock' => $r->stock,
        ], $rows);

        return response()->json(['products' => $products]);
    }
}
